# Step 1: Load and Inspect Raw Expense Data

## Project

Personal Finance Management & Analysis System

## Objective

Load the raw financial transaction dataset and inspect its structure before cleaning.

This notebook will not modify the raw dataset.

In [1]:
import pandas as pd

print("Pandas imported successfully.")

Pandas imported successfully.


In [4]:
from pathlib import Path
import pandas as pd

PROJECT_ROOT = Path.cwd().parent
RAW_FILE = PROJECT_ROOT / "data" / "raw" / "raw_expenses.csv"

print(f"Looking for file at: {RAW_FILE}")
print(f"File exists: {RAW_FILE.exists()}")

Looking for file at: d:\personal-finance-project\data\raw\raw_expenses.csv
File exists: True


In [5]:
df_raw = pd.read_csv(RAW_FILE)

print("Raw dataset loaded successfully.")


Raw dataset loaded successfully.


In [6]:
df_raw.head()


,Date,Account,Category,Subcategory,Note,INR,Income/Expense,Note.1,Amount,Currency
0,3/2/2022 10:11,50.0,Food,NaN,NaN,50.0,Expense,NaN,50.0,INR
1,3/2/2022 10:11,300.0,Other,NaN,NaN,300.0,Expense,NaN,300.0,INR
2,3/1/2022 19:50,78.0,Food,NaN,NaN,78.0,Expense,NaN,78.0,INR
3,3/1/2022 18:56,30.0,Transportation,NaN,NaN,30.0,Expense,NaN,30.0,INR
4,3/1/2022 18:22,67.0,Food,NaN,NaN,67.0,Expense,NaN,67.0,INR


In [7]:
print(df_raw.columns.tolist())


['Date', 'Account', 'Category', 'Subcategory', 'Note', 'INR', 'Income/Expense', 'Note.1', 'Amount', 'Currency']


In [8]:
rows, columns = df_raw.shape

print("Number of rows:", rows)
print("Number of columns:", columns)

Number of rows: 277
Number of columns: 10


In [9]:
df_raw.info()

<class 'pandas.DataFrame'>
RangeIndex: 277 entries, 0 to 276
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            277 non-null    str    
 1   Account         277 non-null    float64
 2   Category        277 non-null    str    
 3   Subcategory     0 non-null      float64
 4   Note            0 non-null      float64
 5   INR             277 non-null    float64
 6   Income/Expense  277 non-null    str    
 7   Note.1          0 non-null      float64
 8   Amount          277 non-null    float64
 9   Currency        277 non-null    str    
dtypes: float64(6), str(4)
memory usage: 21.8 KB


In [10]:
missing_values = df_raw.isna().sum()

missing_values

Date                0
Account             0
Category            0
Subcategory       277
Note              277
INR                 0
Income/Expense      0
Note.1            277
Amount              0
Currency            0
dtype: int64

In [11]:
duplicate_count = df_raw.duplicated().sum()

print("Number of exact duplicate rows:", duplicate_count)

Number of exact duplicate rows: 0


In [12]:
category_counts = df_raw["Category"].value_counts()

category_counts

Category
Food                156
Other                60
Transportation       31
Apparel               7
Household             6
Allowance             6
Social Life           5
Education             1
Salary                1
Self-development      1
Beauty                1
Gift                  1
Petty cash            1
Name: count, dtype: int64

In [13]:
date_preview = pd.to_datetime(
    df_raw["Date"],
    errors="coerce"
)

print("Earliest date:", date_preview.min())
print("Latest date:", date_preview.max())
print("Unparseable dates:", date_preview.isna().sum())

Earliest date: 2021-11-21 13:30:00
Latest date: 2022-03-02 10:11:00
Unparseable dates: 0


## Initial Observations

- The dataset contains 277 rows and 10 columns.
- The transaction dates range from 2021-11-21 to 2022-03-02.
- All dates can be parsed successfully.
- The `Subcategory`, `Note`, and `Note.1` columns are completely empty.
- The `Account` column contains no missing values but requires further inspection.
- The `INR` and `Amount` columns contain numeric values.
- There are no exact duplicate rows.
- The largest categories are Food, Other, and Transportation.
- The raw dataset contains categories that must later be mapped to the project taxonomy.
- The `Salary` category represents income and should not remain an expense category.

### Next Step

In the next notebook step, we will inspect the `Account`, `INR`, `Amount`, and `Income/Expense` columns before cleaning them.

In [14]:
import sys
import pandas as pd

print("Python environment:")
print(sys.executable)

print("\nPandas version:")
print(pd.__version__)

Python environment:
d:\personal-finance-project\.venv\Scripts\python.exe

Pandas version:
3.0.6


In [15]:
print("Shape:", df_raw.shape)
print("\nColumns:")
print(df_raw.columns.tolist())

print("\nData types:")
print(df_raw.dtypes)

print("\nMissing values:")
print(df_raw.isna().sum())

print("\nFirst five rows:")
display(df_raw.head())

Shape: (277, 10)

Columns:
['Date', 'Account', 'Category', 'Subcategory', 'Note', 'INR', 'Income/Expense', 'Note.1', 'Amount', 'Currency']

Data types:
Date                  str
Account           float64
Category              str
Subcategory       float64
Note              float64
INR               float64
Income/Expense        str
Note.1            float64
Amount            float64
Currency              str
dtype: object

Missing values:
Date                0
Account             0
Category            0
Subcategory       277
Note              277
INR                 0
Income/Expense      0
Note.1            277
Amount              0
Currency            0
dtype: int64

First five rows:


,Date,Account,Category,Subcategory,Note,INR,Income/Expense,Note.1,Amount,Currency
0,3/2/2022 10:11,50.0,Food,NaN,NaN,50.0,Expense,NaN,50.0,INR
1,3/2/2022 10:11,300.0,Other,NaN,NaN,300.0,Expense,NaN,300.0,INR
2,3/1/2022 19:50,78.0,Food,NaN,NaN,78.0,Expense,NaN,78.0,INR
3,3/1/2022 18:56,30.0,Transportation,NaN,NaN,30.0,Expense,NaN,30.0,INR
4,3/1/2022 18:22,67.0,Food,NaN,NaN,67.0,Expense,NaN,67.0,INR


In [16]:
display(df_raw.head().style)

,Date,Account,Category,Subcategory,Note,INR,Income/Expense,Note.1,Amount,Currency
0,3/2/2022 10:11,50.000000,Food,nan,nan,50.000000,Expense,nan,50.000000,INR
1,3/2/2022 10:11,300.000000,Other,nan,nan,300.000000,Expense,nan,300.000000,INR
2,3/1/2022 19:50,78.000000,Food,nan,nan,78.000000,Expense,nan,78.000000,INR
3,3/1/2022 18:56,30.000000,Transportation,nan,nan,30.000000,Expense,nan,30.000000,INR
4,3/1/2022 18:22,67.000000,Food,nan,nan,67.000000,Expense,nan,67.000000,INR


In [17]:
print(df_raw.head().to_string())
print("\nUnique currencies:")
print(df_raw["Currency"].unique())

print("\nIncome/Expense values:")
print(df_raw["Income/Expense"].value_counts(dropna=False))

print("\nDate examples:")
print(df_raw["Date"].head(10).tolist())

             Date  Account        Category  Subcategory  Note    INR Income/Expense  Note.1  Amount Currency
0  3/2/2022 10:11     50.0            Food          NaN   NaN   50.0        Expense     NaN    50.0      INR
1  3/2/2022 10:11    300.0           Other          NaN   NaN  300.0        Expense     NaN   300.0      INR
2  3/1/2022 19:50     78.0            Food          NaN   NaN   78.0        Expense     NaN    78.0      INR
3  3/1/2022 18:56     30.0  Transportation          NaN   NaN   30.0        Expense     NaN    30.0      INR
4  3/1/2022 18:22     67.0            Food          NaN   NaN   67.0        Expense     NaN    67.0      INR

Unique currencies:
<StringArray>
['INR', 'USD']
Length: 2, dtype: str

Income/Expense values:
Income/Expense
Expense    231
Income      46
Name: count, dtype: int64

Date examples:
['3/2/2022 10:11', '3/2/2022 10:11', '3/1/2022 19:50', '3/1/2022 18:56', '3/1/2022 18:22', '3/1/2022 18:22', '3/1/2022 18:21', '2/28/2022 11:56', '2/28/2022 11:45',

In [19]:
df = df_raw.copy()

empty_columns = df.columns[df.isna().all()].tolist()
df = df.drop(columns=empty_columns)

df["Date"] = pd.to_datetime(
    df["Date"].astype("string").str.strip(),
    format="mixed",
    dayfirst=True,
    errors="coerce"
)

print("Removed empty columns:", empty_columns)
print("Cleaned shape:", df.shape)
print("Invalid dates:", df["Date"].isna().sum())
print("\nDate range:")
print(df["Date"].min(), "to", df["Date"].max())

display(df.head())

Removed empty columns: ['Subcategory', 'Note', 'Note.1']
Cleaned shape: (277, 7)
Invalid dates: 0

Date range:
2021-01-12 14:00:00 to 2022-12-02 21:20:00


,Date,Account,Category,INR,Income/Expense,Amount,Currency
0,2022-02-03 10:11:00,50.0,Food,50.0,Expense,50.0,INR
1,2022-02-03 10:11:00,300.0,Other,300.0,Expense,300.0,INR
2,2022-01-03 19:50:00,78.0,Food,78.0,Expense,78.0,INR
3,2022-01-03 18:56:00,30.0,Transportation,30.0,Expense,30.0,INR
4,2022-01-03 18:22:00,67.0,Food,67.0,Expense,67.0,INR


In [20]:
print("Duplicate rows:", df.duplicated().sum())

print("\nMissing values:")
print(df.isna().sum())

print("\nNegative amounts:")
print((df["Amount"] < 0).sum())

print("\nIncome/Expense values:")
print(df["Income/Expense"].value_counts(dropna=False))

print("\nCurrencies:")
print(df["Currency"].value_counts(dropna=False))

print("\nAccounts:")
print(df["Account"].value_counts(dropna=False))

print("\nCategories:")
print(df["Category"].value_counts(dropna=False))

Duplicate rows: 0

Missing values:
Date              0
Account           0
Category          0
INR               0
Income/Expense    0
Amount            0
Currency          0
dtype: int64

Negative amounts:
0

Income/Expense values:
Income/Expense
Expense    231
Income      46
Name: count, dtype: int64

Currencies:
Currency
INR    276
USD      1
Name: count, dtype: int64

Accounts:
Account
40.0     14
200.0    11
25.0     10
30.0      9
100.0     9
         ..
95.0      1
179.0     1
90.0      1
97.0      1
875.0     1
Name: count, Length: 137, dtype: int64

Categories:
Category
Food                156
Other                60
Transportation       31
Apparel               7
Household             6
Allowance             6
Social Life           5
Education             1
Salary                1
Self-development      1
Beauty                1
Gift                  1
Petty cash            1
Name: count, dtype: int64


In [21]:
print("Accounts:")
print(df["Account"].value_counts(dropna=False).to_string())

print("\nCategories:")
print(df["Category"].value_counts(dropna=False).to_string())

Accounts:
Account
40.00       14
200.00      11
25.00       10
30.00        9
100.00       9
20.00        9
50.00        8
80.00        8
1000.00      8
300.00       6
70.00        6
150.00       6
500.00       5
15.00        5
1500.00      4
120.00       4
10.00        4
250.00       3
60.00        3
43.00        3
115.00       3
2000.00      3
110.00       3
75.00        3
130.00       3
399.00       2
138.00       2
125.00       2
18.00        2
340.00       2
240.00       2
106.00       2
8000.00      2
140.00       2
85.00        2
504.00       2
400.00       2
160.00       2
65.00        2
58.00        2
78.00        1
67.00        1
339.15       1
243.00       1
1700.00      1
84.00        1
33.00        1
303.00       1
289.00       1
324.70       1
83.00        1
1300.00      1
336.00       1
201.80       1
111.00       1
1575.00      1
380.00       1
270.00       1
348.00       1
165.00       1
475.00       1
491.00       1
10000.00     1
373.00       1
641.00       1
2099.00

In [22]:
print("Non-INR transactions:")
display(df[df["Currency"] != "INR"])

Non-INR transactions:


,Date,Account,Category,INR,Income/Expense,Amount,Currency
217,2021-08-12 14:15:00,15.0,Food,1120.72,Expense,15.0,USD


In [23]:
print("INR column summary:")
print(df["INR"].describe())

print("\nAmount column summary:")
print(df["Amount"].describe())

INR column summary:
count      277.000000
mean       410.750903
std       1065.756569
min          3.000000
25%         50.000000
50%        128.000000
75%        301.150000
max      10000.000000
Name: INR, dtype: float64

Amount column summary:
count      277.000000
mean       406.759134
std       1065.158318
min          3.000000
25%         50.000000
50%        125.000000
75%        300.000000
max      10000.000000
Name: Amount, dtype: float64


In [25]:
df["Amount_INR"] = df["INR"]

print(df[[
    "Date",
    "Category",
    "Income/Expense",
    "Amount",
    "Currency",
    "Amount_INR"
]].head())

                 Date        Category Income/Expense  Amount Currency  \
0 2022-02-03 10:11:00            Food        Expense    50.0      INR   
1 2022-02-03 10:11:00           Other        Expense   300.0      INR   
2 2022-01-03 19:50:00            Food        Expense    78.0      INR   
3 2022-01-03 18:56:00  Transportation        Expense    30.0      INR   
4 2022-01-03 18:22:00            Food        Expense    67.0      INR   

   Amount_INR  
0        50.0  
1       300.0  
2        78.0  
3        30.0  
4        67.0  


In [26]:
income_total = df.loc[
    df["Income/Expense"] == "Income",
    "Amount_INR"
].sum()

expense_total = df.loc[
    df["Income/Expense"] == "Expense",
    "Amount_INR"
].sum()

net_total = income_total - expense_total

print(f"Total income: ₹{income_total:,.2f}")
print(f"Total expenses: ₹{expense_total:,.2f}")
print(f"Net amount: ₹{net_total:,.2f}")

Total income: ₹54,754.00
Total expenses: ₹59,024.00
Net amount: ₹-4,270.00


In [27]:
category_summary = (
    df[df["Income/Expense"] == "Expense"]
    .groupby("Category", as_index=False)
    .agg(
        Total_INR=("Amount_INR", "sum"),
        Transactions=("Amount_INR", "count"),
        Average_INR=("Amount_INR", "mean")
    )
    .sort_values("Total_INR", ascending=False)
)

category_summary["Percentage_of_Expenses"] = (
    category_summary["Total_INR"]
    / category_summary["Total_INR"].sum()
    * 100
)

display(category_summary)

,Category,Total_INR,Transactions,Average_INR,Percentage_of_Expenses
3,Food,24502.48,156,157.067179,41.512741
5,Household,12188.00,6,2031.333333,20.649227
9,Transportation,9203.80,31,296.896774,15.593318
6,Other,5117.00,22,232.590909,8.669355
0,Apparel,3388.00,7,484.000000,5.740038
8,Social Life,2513.72,5,502.744000,4.258810
2,Education,1400.00,1,1400.000000,2.371917
7,Self-development,400.00,1,400.000000,0.677690
1,Beauty,196.00,1,196.000000,0.332068
4,Gift,115.00,1,115.000000,0.194836


In [28]:
display(category_summary.head(5))

,Category,Total_INR,Transactions,Average_INR,Percentage_of_Expenses
3,Food,24502.48,156,157.067179,41.512741
5,Household,12188.00,6,2031.333333,20.649227
9,Transportation,9203.80,31,296.896774,15.593318
6,Other,5117.00,22,232.590909,8.669355
0,Apparel,3388.00,7,484.000000,5.740038


In [29]:
monthly_summary = (
    df.assign(
        Month=df["Date"].dt.to_period("M").astype(str)
    )
    .groupby(["Month", "Income/Expense"], as_index=False)
    .agg(
        Total_INR=("Amount_INR", "sum"),
        Transactions=("Amount_INR", "count")
    )
)

monthly_pivot = (
    monthly_summary
    .pivot(
        index="Month",
        columns="Income/Expense",
        values="Total_INR"
    )
    .fillna(0)
    .reset_index()
)

monthly_pivot["Net_INR"] = (
    monthly_pivot.get("Income", 0)
    - monthly_pivot.get("Expense", 0)
)

display(monthly_pivot)

Income/Expense,Month,Expense,Income,Net_INR
0,2021-01,356.00,0.0,-356.00
1,2021-02,579.00,0.0,-579.00
2,2021-03,215.00,1000.0,785.00
3,2021-04,619.00,0.0,-619.00
4,2021-05,6625.00,8000.0,1375.00
5,2021-06,815.00,0.0,-815.00
6,2021-07,40.00,0.0,-40.00
7,2021-08,1230.72,0.0,-1230.72
8,2021-09,455.00,0.0,-455.00
9,2021-10,436.00,0.0,-436.00


In [30]:
all_months = pd.period_range(
    start=df["Date"].min().to_period("M"),
    end=df["Date"].max().to_period("M"),
    freq="M"
).astype(str)

monthly_pivot = (
    monthly_pivot
    .set_index("Month")
    .reindex(all_months, fill_value=0)
    .rename_axis("Month")
    .reset_index()
)

for column in ["Income", "Expense"]:
    if column not in monthly_pivot.columns:
        monthly_pivot[column] = 0.0

monthly_pivot["Net_INR"] = (
    monthly_pivot["Income"] - monthly_pivot["Expense"]
)

display(monthly_pivot)

Income/Expense,Month,Expense,Income,Net_INR
0,2021-01,356.00,0.0,-356.00
1,2021-02,579.00,0.0,-579.00
2,2021-03,215.00,1000.0,785.00
3,2021-04,619.00,0.0,-619.00
4,2021-05,6625.00,8000.0,1375.00
5,2021-06,815.00,0.0,-815.00
6,2021-07,40.00,0.0,-40.00
7,2021-08,1230.72,0.0,-1230.72
8,2021-09,455.00,0.0,-455.00
9,2021-10,436.00,0.0,-436.00


In [31]:
REPORTS_DIR = PROJECT_ROOT / "reports"
REPORTS_DIR.mkdir(exist_ok=True)

monthly_pivot.to_csv(
    REPORTS_DIR / "monthly_summary.csv",
    index=False
)

category_summary.to_csv(
    REPORTS_DIR / "category_summary.csv",
    index=False
)

print("Saved monthly_summary.csv and category_summary.csv")

Saved monthly_summary.csv and category_summary.csv
